handler/aggregate/loader.go

```diff
@@ -42,6 +41,21 @@ func (l *Loader) Load(
 		return nil, err
 	}
 
+	ss, ok, err := l.AggregateRepository.LoadAggregateSnapshot(ctx, hk, id)
+	if err != nil {
+		return nil, err
+	}
+
+	if ok {
+		s, err := l.Marshaler.Unmarshal(ss.Packet)
+		if err != nil {
+			return nil, err
+		}
+
+		base = s.(dogma.AggregateRoot)
+		md.BarrierEventID = ss.Version
```

Comment:
There's a subtle issue with setting `md.BarrierEventID` here. Doing so makes `l.applyEvents()` behave as desired but it will also be persisted in the `AggregateMetaData` on the next save which essentially marks the instance as destroyed as of that event.

This probably has no observable effect so long as the snapshot always remains available, but there's one particular case where snapshots may need to be removed, which is when the `AggregateRoot` implementation is changed to the point where old snapshots are no longer valid representation.

The "standard" approach in this scenario is to just delete the incompatible snapshots so that new snapshots are built from the event history, but because `BarrierEventID` has been set these new snapshots will not "see" the entire event history.

I would suggest adding an additional `barrierEventID` parameter to `applyEvents()`, pass it `md.BarrierEventID` by default, but override it with `ss.Version` if available, and avoid modifying `md.BarrierEventID` altogether.